### PR TITLE
WordPress.com Push Authentication

### LINKED ISSUE #3615: WordPress.com Push Authentication
#### Description:
Implement a 'One Tap' mechanism to ease the WordPress.com Login, when Two Factor Auth is enabled.

#### Details:
Whenever a dotcom user attempts to login, if his account has Two Factor enabled, he'll be required to enter a password.

As an alternative to simply typing the One Time Code, let's add support for a new mechanism that would allow users to verify Web Logins Attempts, using their WPiOS Installs running on their devices.

In the same WordPress.com Login Form in which they'd normally enter the code, we'll render a link that, on click, would post a Push Notification to the device. When received, the app should render an UI explaining what's going on, and should allow the users to verify the login attempt with a single tap.

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Description:
The goal of this Pull Request is to allow WordPress.com users *-whose accounts have Two Factor Authentication Enabled-* to easily authorize online access, by means of the WordPress iOS App.

#### Steps:
1. Fresh install WPiOS in the device
2. Log into your WordPress.com account
3. Open WordPress.com in a browser
4. Attempt to log into an account with 2FA enabled (a8c only)

#### Expected:
The app should get a Push Notification indicating there's been a login attempt, and it should allow the user to Authorize / Ignore.

This will be live only for WPiOS 5.3+, so, please, make sure that the Version String is set to 5.3.

Closes #3615

### CODE REVIEW

**Comment by diegoreymendez** [comment]
@sendhil, @jleandroperez - I'm just going through the list of PRs and noticed this one is tagged as "not ready for review".  Is that still the case?

**Comment by jleandroperez** [comment]
@diegoreymendez iOS code is ready, but we're hanging out until the backend's code is merged. Will keep you posted, thanks for checking Diego!

**Comment by jleandroperez** [comment]
@diegoreymendez ready for review. Please, poke me whenever you get to it.

Thank you!

**Comment by jleandroperez** [comment]
I may have (accidentally) flooded @diegoreymendez with PR's (sorry about that Diego!).

@aerych may i bother you with a review instead?.

**Many many thanks in advance!**

File: WordPress/Classes/Networking/PushAuthenticationServiceRemote.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            return nil
+        }
+    }
+    
+
+    /**
+    *  @details     Verifies a WordPress.com Login.
+    *  @param       token       The token passed on by WordPress.com's 2FA Push Notification.
+    *  @param       success     Closure to be executed on success. CAn be nil.
+    *  @param       failure     Closure to be executed on failure. CAn be nil.
```
Comment (aerych) [inline WordPress/Classes/Networking/PushAuthenticationServiceRemote.swift]: CAn

File: WordPress/Classes/Services/PushAuthenticationService.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            return
+        }
+        
+        self.authenticationServiceRemote!.authorizeLogin(token,
+            success:    {
+                            completion(true)
+                        },
+            failure:    {
+                            completion(false)
+                        })
```
Comment (aerych) [inline WordPress/Classes/Services/PushAuthenticationService.swift]: The formatting looks a bit odd but I think I'm getting used to it. :)

**Comment by aerych** [comment]
Reviewed and took for a spin. :)   Code looks good @jleandroperez and I only had one really nitpicky comment. 

One thing that was a bit weird (maybe because of being on a develop build?) was the formatting of the push notification and the prompt.  I was expecting to see either my username, my name, or my email addy as identifying who was trying to login in but I'm just seeing dashes.  Examples:

notification
prompt

That seems more like an issue that's server-side.  If you can confirm then I think the rest of the PR is good to go.   :shipit:

File: WordPress/Classes/Networking/PushAuthenticationServiceRemote.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/Networking/PushAuthenticationServiceRemote.swift]: Oops, fixing! thanks!

File: WordPress/Classes/Services/PushAuthenticationService.swift (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/Services/PushAuthenticationService.swift]: I'm still trying to figure out the best way to format this. (Imho) it's easier to tell the beginning / ending of a closure param this way. But maybe the alternative is better?

```
success:    {
    completion(true)
},
failure:    {
completion(false)
})
```

**Comment by jleandroperez** [comment]
Verified, the AlertView text comes from the backend. Android is experiencing the same glitch.
I'll update this thread as soon as i've got backend news.

@aerych thanks for the review!

**Pull request merged**